pkg/cli/admin/nodeimage/create.go

```diff
@@ -0,0 +1,578 @@
+				if !errors.As(err, &codeExitErr) {
+					return false, err
+				}
+				if codeExitErr.Code != 1 {
+					return false, fmt.Errorf("unexpected error code: %w", codeExitErr)
+				}
+				return false, nil
+			}
+
+			// Extract node-joiner tool exit code on completion
+			o.nodeJoinerExitCode, err = strconv.Atoi(w.String())
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		})
+}
+
+func (o *CreateOptions) runNodeJoinerPod(ctx context.Context) error {
+	tasks := []func(context.Context) error{
+		o.getNodeJoinerPullSpec,
+		o.createNamespace,
+		o.createServiceAccount,
+		o.createRolesAndBindings,
+		o.createInputConfigMap,
+		o.createPod,
+	}
+	for _, task := range tasks {
+		if err := task(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (o *CreateOptions) getNodeJoinerPullSpec(ctx context.Context) error {
+	// Get the current cluster release version.
+	releaseImage, err := o.fetchClusterReleaseImage(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Extract the baremetal-installer image pullspec, since it
+	// provide the node-joiner tool.
+	opts := ocrelease.NewInfoOptions(o.IOStreams)
+	opts.SecurityOptions = o.SecurityOptions
+	release, err := opts.LoadReleaseInfo(releaseImage, false)
+	if err != nil {
+		return err
+	}
+
+	tagName := "baremetal-installer"
+	for _, tag := range release.References.Spec.Tags {
+		if tag.Name == tagName {
+			o.nodeJoinerImage = tag.From.Name
+			return nil
+		}
+	}
+
+	return fmt.Errorf("no image tag %q exists in the release image %s", tagName, releaseImage)
+}
+
+func (o *CreateOptions) fetchClusterReleaseImage(ctx context.Context) (string, error) {
+	cv, err := o.ConfigClient.ConfigV1().ClusterVersions().Get(ctx, "version", metav1.GetOptions{})
+	if err != nil {
+		if kapierrors.IsNotFound(err) || kapierrors.ReasonForError(err) == metav1.StatusReasonUnknown {
+			klog.V(2).Infof("Unable to find cluster version object from cluster: %v", err)
+			return "", fmt.Errorf("command expects a connection to an OpenShift 4.x server")
+		}
+	}
+	image := cv.Status.Desired.Image
+	if len(image) == 0 && cv.Spec.DesiredUpdate != nil {
+		image = cv.Spec.DesiredUpdate.Image
+	}
+	if len(image) == 0 {
+		return "", fmt.Errorf("the server is not reporting a release image at this time")
+	}
+
+	return image, nil
+}
+
+func (o *CreateOptions) createNamespace(ctx context.Context) error {
+	nsNodeJoiner := &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "openshift-node-joiner-",
+			Annotations: map[string]string{
+				"oc.openshift.io/command":    "oc adm node-image create",
+				"openshift.io/node-selector": "",
+			},
+		},
+	}
+
+	ns, err := o.Client.CoreV1().Namespaces().Create(ctx, nsNodeJoiner, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("cannot create namespace: %w", err)
+	}
+
+	o.nodeJoinerNamespace = ns
+	return nil
+}
+
+func (o *CreateOptions) cleanup(ctx context.Context) {
+	if o.nodeJoinerNamespace == nil {
+		return
```

Comment:
How can this be possible. I'm asking to ensure that if this line executes, we won't have any leaked resources. Because customers complain about leftover clusterroles, serviceaccounts, etc.?